crates/ruff_python_semantic/src/analyze/class.rs

```diff
@@ -100,23 +97,47 @@ pub fn any_super_class(
 
 /// Return `true` if `class_def` is a class that has one or more enum classes in its mro
 pub fn is_enumeration(class_def: &ast::StmtClassDef, semantic: &SemanticModel) -> bool {
-    any_qualified_name(class_def, semantic, &|qualified_name| {
-        matches!(
-            qualified_name.segments(),
-            [
-                "enum",
-                "Enum" | "Flag" | "IntEnum" | "IntFlag" | "StrEnum" | "ReprEnum" | "CheckEnum"
-            ]
-        )
+    any_qualified_name(class_def, semantic, &|expr| {
+        semantic
+            .resolve_qualified_name(map_subscript(expr))
+            .is_some_and(|qualified_name| {
+                matches!(
+                    qualified_name.segments(),
+                    [
+                        "enum",
+                        "Enum"
+                            | "Flag"
+                            | "IntEnum"
+                            | "IntFlag"
+                            | "StrEnum"
+                            | "ReprEnum"
+                            | "CheckEnum"
+                    ]
+                )
+            })
     })
 }
 
 /// Returns `true` if the given class is a metaclass.
 pub fn is_metaclass(class_def: &ast::StmtClassDef, semantic: &SemanticModel) -> bool {
-    any_qualified_name(class_def, semantic, &|qualified_name| {
-        matches!(
-            qualified_name.segments(),
-            ["" | "builtins", "type"] | ["abc", "ABCMeta"] | ["enum", "EnumMeta" | "EnumType"]
-        )
+    any_qualified_name(class_def, semantic, &|expr| match expr {
+        Expr::Call(ast::ExprCall { func, .. }) => {
+            // Ex) `class Foo(type(Protocol)): ...`
+            semantic
+                .resolve_qualified_name(func.as_ref())
+                .is_some_and(|qualified_name| {
+                    matches!(qualified_name.segments(), ["" | "builtins", "type"])
```

Comment:
Are any of the others allowed as calls? I don't think so.

Reply:
Not in the same way... But you need to distinguish between the two different ways of calling type here. X is a metaclass but Y is not:

```pycon
>>> class X(type(int)): pass
... 
>>> class Y(type('Foo', (), {})): pass
... 
>>> class A(metaclass=X): pass
... 
>>> class B(metaclass=Y): pass
... 
Traceback (most recent call last):
  File "<string>", line 1, in <module>
TypeError: Y() takes no arguments
```

Reply:
Not sure this has been addressed. A `type()` call is only likely to create a metaclass if it's the one-argument `type()` call -- the three-argument `type()` call is just going to create an instance of `type` (i.e., a normal class). So I think we should do something like this:

```diff
diff --git a/crates/ruff_python_semantic/src/analyze/class.rs b/crates/ruff_python_semantic/src/analyze/class.rs
index 73ab1c58a..7277fad21 100644
--- a/crates/ruff_python_semantic/src/analyze/class.rs
+++ b/crates/ruff_python_semantic/src/analyze/class.rs
@@ -124,13 +124,16 @@ pub fn is_enumeration(class_def: &ast::StmtClassDef, semantic: &SemanticModel) -
 /// Returns `true` if the given class is a metaclass.
 pub fn is_metaclass(class_def: &ast::StmtClassDef, semantic: &SemanticModel) -> bool {
     any_base_class(class_def, semantic, &|expr| match expr {
-        Expr::Call(ast::ExprCall { func, .. }) => {
+        Expr::Call(ast::ExprCall {
+            func, arguments, ..
+        }) => {
             // Ex) `class Foo(type(Protocol)): ...`
-            semantic
-                .resolve_qualified_name(func.as_ref())
-                .is_some_and(|qualified_name| {
-                    matches!(qualified_name.segments(), ["" | "builtins", "type"])
-                })
+            arguments.len() == 1
+                && semantic
+                    .resolve_qualified_name(func.as_ref())
+                    .is_some_and(|qualified_name| {
+                        matches!(qualified_name.segments(), ["" | "builtins", "type"])
+                    })
         }
```

Reply:
I considered this not worth supporting but it's simple enough if you prefer!